mipengine/controller/algorithm_flow_data_objects.py

```diff
@@ -100,60 +115,71 @@ def get_table_data(self) -> List[Union[int, float, str]]:
         return merged_table_data
 
     def __repr__(self):
-        r = f"\n\tLocalNodeTable: {self.get_table_schema()}\n"
-        for node, table_name in self.nodes_tables.items():
+        tables_schema = next(
+            iter(self.nodes_tables_info.values())
+        ).schema_  # Get any schema from the tables
+        r = f"\n\tLocalNodeTable: {tables_schema=}\n"
+        for node, table_name in self.nodes_tables_info.items():
             r += f"\t{node=} {table_name=}\n"
         return r
 
-    def _validate_matching_table_names(self, table_names: List[TableName]):
-        table_name_without_node_id = table_names[0].without_node_id()
-        for table_name in table_names:
-            if table_name.without_node_id() != table_name_without_node_id:
+    def _validate_matching_table_names(self):
+        table_infos = list(self._nodes_tables_info.values())
+        table_name_without_node_id = table_infos[0].name_without_node_id
+        for table_name in table_infos:
+            if table_name.name_without_node_id != table_name_without_node_id:
                 raise MismatchingTableNamesException(
-                    [table_name.full_table_name for table_name in table_names]
+                    [table_info.name for table_info in table_infos]
                 )
 
 
 class GlobalNodeTable(GlobalNodeData):
     _node: GlobalNode
-    _table: TableName
+    _table_info: TableInfo
 
-    def __init__(self, node: GlobalNode, table: TableName):
+    def __init__(self, node: GlobalNode, table_info: TableInfo):
         self._node = node
-        self._table = table
+        self._table_info = table_info
+
+    @property
+    def schema(self):
+        return self._table_info.schema_
+
+    @property
+    def columns(self):
+        return [
+            (column.name, column.dtype) for column in self._table_info.schema_.columns
+        ]
 
     @property
     def node(self) -> GlobalNode:
         return self._node
 
     @property
-    def table(self) -> TableName:
-        return self._table
-
-    # TODO this is redundant, either remove it or overload all node methods here?
-    def get_table_schema(self) -> TableSchema:
-        return self._node.get_table_schema(self.table)
+    def table_info(self) -> TableInfo:
```

Comment:
The `TableInfo` is used in various places mainly to get the name/schema from it.

I didn't want to create methods returning each part of the `TableInfo` because it could change in the future and I didn't see the need for it. Also in the `LocalNodesTable` you have a dict of `TableInfo` objects which is used as is. I though we should keep the logic similar.

What do you think?

Reply:
what I m trying to say is that there is some kind of duplication in the functionality here because the schema of the table(s) can be acquired in 2 ways (almost 3 ways, if you also count the `columns` property), either from the `schema` property or from within the `InfoTable` object that is returned from the `nodes_tables_info` property. Having multiple ways of doing the same thing is confusing. Also, although I m not entirely sure what "could change in the future", if you mean the `TableInfo` class, then that is one more reason to fully encapsulate(hide) the `TableInfo` class from the user of the `GlobalNodeTable` and `LocalNodesTable` classes so no matter how `TableInfo` is changed, the changes do not propagate further. But anyhow it is not very important..

Reply:
Personally, since the `TableInfo` is an immutable DTO I would return it whole as a property.

So in the algorithm flow you would need to do `X.table_info.schema`  or `X.table_info.columns`

But since Jason asked for something even more simple: `X.columns` that's where the confusion probably arises.

My point for future changes is that if any property is added in the `TableInfo`, we would need to keep in mind that we must also add a new method in the algorithm flow objects.

So what do you think we should do here?